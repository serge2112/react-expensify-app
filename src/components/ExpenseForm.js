import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';


export default class ExpenseForm extends React.Component{
	constructor(props){
		super(props);
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100): '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			createdAtFocused: false,
			error: ''
		}
	};
	
	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({description}));
	};

	onNoteChange = (e) => {
		const note = e.target.value;
		this.setState(() => ({note}))
	};

	onAmountChange = (e) => {
		const amount = e.target.value;

		if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
			this.setState(() => ({amount}));
		}
	};

	onDateChanged = (createdAt) => {
		if(createdAt){
			this.setState(() => ({createdAt}));
		}		
	};

	onFocusChanged = (focused) => {
		this.setState(() => ({createdAtFocused: focused}));
	};

	onFormSubmit = (e) => {
		e.preventDefault();
		if(!this.state.description || !this.state.amount){
			this.setState(() => ({error: 'please provide a description and an amount'}));
		}else{
			this.setState(() => ({error: ''}));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};

	render(){
		return (				
				<form className='form' onSubmit={this.onFormSubmit}>
					<div className='form__error'>
						{this.state.error && <p>{this.state.error}</p>} 
					</div>					
					<input type='text'
						className='text-input'
						placeholder='description'
						autoFocus 
						value={this.state.description}
						onChange={this.onDescriptionChange}/>
					<input type='text' 
						className='text-input'
						placeholder='amount'
						value={this.state.amount}
						onChange={this.onAmountChange}/>
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChanged}
						focused={this.state.createdAtFocused}
						onFocusChange={this.onFocusChanged}
						id="createdAt" 
						numberOfMonths={1}
						isOutsideRange={() => {false}}/>					
					<textarea placeholder='type a note for your expense (optional)'
						className='textarea'
						value={this.state.note}
						onChange={this.onNoteChange}>
					</textarea>
					<div>
						<button className='button'>Add Expense</button>
					</div>					
				</form>			
		)		
	}
}