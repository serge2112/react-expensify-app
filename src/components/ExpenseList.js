import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import selectorExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
	<div className='content-container'>
		<div className='list-header'>
			<div className='show-on-mobile'>Expenses</div>
			<div className='show-on-desktop'>Expenses</div>
			<div className='show-on-desktop'>Amount</div>
		</div>	
		<div className='list-body'>
		{
			props.expenses.length === 0 ? 
			(
				<div className='list-item list-item--message'>
					<span>There are not any expenses to show</span>
				</div>				
			) : 
			(				
				props.expenses.map((expense) => { 
					return <ExpenseListItem key={expense.id} {...expense}/>;									
				})
			)
		}
		</div>				
	</div>
);

const mapStateToProps = (state) =>{
	return {
		expenses: selectorExpenses(state.expenses, state.filters)
	}
};

export default connect(mapStateToProps)(ExpenseList);