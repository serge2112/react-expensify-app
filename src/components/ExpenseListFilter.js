import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilter extends React.Component{
	state = {
		calendarFocus: null
	};

	onDateChange = ({startDate, endDate}) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};

	onFocusChange = (calendarFocus) => {
		this.setState(() => ({calendarFocus}))
	};

	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value);
	};

	onSortByChange = (e) => {
		if(e.target.value === 'date'){
			this.props.sortByDate();
		}else if (e.target.value === 'amount'){
			this.props.sortByAmount();
		}
	};

	render() {
		return (
			<div>
				<input type='text' value={this.props.filters.text} onChange={this.onTextChange}/>
				<select value={this.props.filters.sortBy} onChange={this.onSortByChange}>
					<option value='date'>Date</option>
					<option value='amount'>Amount</option>
				</select>
				<DateRangePicker 
					startDate={this.props.filters.startDate}
					startDateId='startDate'					
					endDate={this.props.filters.endDate}
					endDateId='endDate'
					onDatesChange={this.onDateChange}
					focusedInput={this.state.calendarFocus}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => {true}}
					showClearDates={true}
				/>
			</div>
		);
	}
}

const mapStateToProps = (state) => {
	return {
		filters: state.filters
	}
}

const mapDispatchToProps = (dispatch) => {
	return {
		setTextFilter: (text) => dispatch(setTextFilter(text)),
		sortByDate: () => dispatch(sortByDate()),
		sortByAmount: () => dispatch(sortByAmount()),
		setStartDate: (startDate) => dispatch(setStartDate(startDate)),
		setEndDate: (endDate) => dispatch(setEndDate(endDate))
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);