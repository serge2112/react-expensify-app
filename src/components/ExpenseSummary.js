import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectorExpenses from '../selectors/expenses';
import { getExpensesTotal } from '../selectors/expenses-total';
import { Link } from 'react-router-dom';

export const ExpenseSummary = (props) => {
	const expensesWord = props.expenses.length === 1 ? 'expense' : 'expenses';
	const formattedAmount = numeral(getExpensesTotal(props.expenses) / 100).format('$0,0.00');

	return (
		<div className='page-header'>
			<div className="content-container">
				<h1 className="page-header__title">Showing <span>{props.expenses.length}</span> {expensesWord}, adding up <span>{formattedAmount}</span></h1>
				<Link className="button" to="/create">Add Expense</Link>
			</div>			
		</div>
	);
}

const mapStateToProps = (state) => {
	return {
		expenses: selectorExpenses(state.expenses, state.filters)
	}
};

export default connect(mapStateToProps)(ExpenseSummary);