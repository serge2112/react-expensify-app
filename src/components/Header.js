import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogOut } from '../actions/auth';

export const Header = ({ startLogOut }) => (
	<header>
		<h1>Expensify</h1>
		<NavLink to="/dashboard" activeClassName="is-active">Dashboard</NavLink>		
		<NavLink to="/create" activeClassName="is-active">Add Expense</NavLink>		
		<NavLink to="/edit" activeClassName="is-active">Edit Expense</NavLink>		
		<NavLink to="/help" activeClassName="is-active">Help</NavLink>
		<button onClick={startLogOut}>LogOut</button>		
	</header>
);

const mapDispatchToProps = (dispatch) => ({
	startLogOut: () => dispatch(startLogOut())
});

export default connect(undefined, mapDispatchToProps)(Header);
