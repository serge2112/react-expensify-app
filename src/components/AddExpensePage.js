import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component{
	onSubmit = (expense) => {
		this.props.addExpense(expense);
		this.props.history.push('/');
	}
	render() {
		return (
			<div>
				<div>Hello from Add Expense Page</div>
				<ExpenseForm onSubmit={this.onSubmit}></ExpenseForm>
			</div>	
		);
	}
}

const mapDispatchToProps = (dispatch) => (
	{
		addExpense: (expense) => dispatch(addExpense(expense))		
	}
);

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
