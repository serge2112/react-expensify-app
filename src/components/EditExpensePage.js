import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpenses } from '../actions/expenses';

// Refactor EditExpensePage to be a class based component
// Setup mapDispatchToprops editExpense and RemoveExpense

// Should render EditExpensePage
// snapshot

// Should handle editExpense
// Spies

// Should handle removeExpense
// Spies

export class EditExpensePage extends React.Component{
	onSubmit = (expense) => {
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	};
	
	onRemove = () => {
		this.props.startRemoveExpenses({id: this.props.expense.id});
		this.props.history.push('/');
	};

	render(){
		return (
			<div>
				<ExpenseForm 
					expense={this.props.expense}
					onSubmit={this.onSubmit}/>
					<button onClick={this.onRemove}>Remove</button>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => {
	return{
		expense: state.expenses.find(expense => expense.id === props.match.params.id)
	}
};

const mapDispatchToProps = (dispatch, props) => {
	return {
		startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
		startRemoveExpenses: (data) => dispatch(startRemoveExpenses(data))
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
