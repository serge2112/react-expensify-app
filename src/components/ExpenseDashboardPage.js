import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpenseSummary from '../components/ExpenseSummary';

const ExpenseDashboardPage = () => (
	<div>
		<ExpenseListFilter /> 
		<ExpenseSummary />
		<ExpenseList/>
	</div>	
);

export default ExpenseDashboardPage;
