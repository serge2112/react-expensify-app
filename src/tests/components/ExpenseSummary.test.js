import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import {ExpenseSummary} from '../../components/ExpenseSummary';

test('Should show summary with one expense', () => {	
	const wrapper = shallow(<ExpenseSummary expenses={[expenses[0]]}/>);
	expect(wrapper).toMatchSnapshot();
});

test('Should show summary with Multiple expense', () => {	
	const wrapper = shallow(<ExpenseSummary expenses={expenses}/>);
	expect(wrapper).toMatchSnapshot();
});