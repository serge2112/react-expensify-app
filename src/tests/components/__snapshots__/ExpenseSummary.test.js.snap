// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should show summary with Multiple expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Showing 
      <span>
        3
      </span>
       
      expenses
      , adding up 
      <span>
        $1,807.90
      </span>
    </h1>
    <Link
      className="button"
      to="/create"
    >
      Add Expense
    </Link>
  </div>
</div>
`;

exports[`Should show summary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Showing 
      <span>
        1
      </span>
       
      expense
      , adding up 
      <span>
        $1.95
      </span>
    </h1>
    <Link
      className="button"
      to="/create"
    >
      Add Expense
    </Link>
  </div>
</div>
`;
